### PR TITLE
Adapt GridLayout to work with UI

### PR DESCRIPTION
**Pull request by antrikshmisri**

### CODE REVIEW

**Comment by antrikshmisri** [comment]
Hey @skoudoro, I have created a rough example for layout with UI elements, which is present in `docs/tutorials/02_ui/viz_layout.py`. What else needs to be done in this PR?

**Comment by skoudoro** [comment]
Hi @antrikshmisri,

Thank you for this PR and there is no more to do for now. I mean until the UI redesign PR is not in.

**Comment by skoudoro** [comment]
UI refactoring is IN!  Can you rebase this PR and update the import. Thank you

**Comment by antrikshmisri** [comment]
> UI refactoring is IN! Can you rebase this PR and update the import. Thank you

Hey @skoudoro, that's greate news. But when trying to import `ui` into `utils` module using `from fury import ui`, I am still getting the following error:-

```python 
  File "D:\pythonProjects\fury-event\fury\fury\utils.py", line 6, in <module>      
    from fury import ui
  File "D:\pythonProjects\fury-event\fury\fury\ui\__init__.py", line 2, in <module>
    from fury.ui.core import *
  File "D:\pythonProjects\fury-event\fury\fury\ui\core.py", line 12, in <module>
    from fury.utils import set_input
ImportError: cannot import name 'set_input'
```
Am I doing something wrong?

File: fury/utils.py [hunk 1]
```diff
@@ -753,6 +753,12 @@ def numpy_to_vtk_matrix(array):
 
 def get_bounding_box_sizes(actor):
     """Get the bounding box sizes of an actor."""
+    from fury.ui import UI  # to avoid circular import
+
+    if isinstance(actor, UI):
+        width, height = actor.size
+        return (width, height, 0)
```
Comment (skoudoro) [inline fury/utils.py:760]: I recommend avoiding importing `UI` in `utils`. `Utils` is dedicated to `actor`. 

I understand that you want only one function to get the size so you should define this new function in the layout class

**Review by skoudoro — COMMENTED** [review]
see below my comment

File: fury/utils.py (hunk 1 above)
Comment (antrikshmisri) [reply fury/utils.py]: Hey @skoudoro, thanks for the suggestion/reminder. I have removed the size calculation from `utils` and added it to `GridLayout` class. Unfortunately, importing `UI` in `layout` module by `from fury.ui.core import UI` doesn't work either for me. I get the following error:-
```python
File "D:\pythonProjects\fury-event\fury\fury\layout.py", line 5, in <module>        
    from fury.ui.core import UI
  File "D:\pythonProjects\fury-event\fury\fury\ui\__init__.py", line 3, in <module>   
    from fury.ui.containers import *
  File "D:\pythonProjects\fury-event\fury\fury\ui\containers.py", line 12, in <module>
    from fury.actor import grid
  File "D:\pythonProjects\fury-event\fury\fury\actor.py", line 2135, in <module>
    class Container(object):
  File "D:\pythonProjects\fury-event\fury\fury\actor.py", line 2152, in Container
    def __init__(self, layout=layout.Layout()):
AttributeError: module 'fury.layout' has no attribute 'Layout'
```

**Comment by antrikshmisri** [comment]
Hey @skoudoro, the error has been fixed by implementing the logic as discussed in the meeting. Thanks!

**Review by skoudoro — COMMENTED** [review]
Thank you for the update @antrikshmisri. 

Please, could you check the CI  to see why they are all failing? Something seems wrong with `containers`.

Thanks

**Comment by antrikshmisri** [comment]
> Please, could you check the CI to see why they are all failing? Something seems wrong with `containers`.

Hey @skoudoro, the problem is that the `Container` class has a method called `add_to_scene` which means it will be considered as a UI element instead of an actor in `get_bounding_box_sizes` method. I think it would be nice to compute the sizes of just actors in `get_bounding_box_sizes`.

File: fury/layout.py [hunk 2]
```diff
@@ -119,4 +135,21 @@ def compute_positions(self, actors):
...
         return positions
+
+    def compute_sizes(self, actor):
+        """Compute the bounding box size of the actor/UI element
+
+        Parameters
+        ---------
+        actor: `vtkProp3D` or `UI` element
+            Actor/UI element whose size is to be calculated
+        """
```
Comment (skoudoro) [inline fury/layout.py:149]: do not forget to add the `Returns`in the docstring

File: fury/utils.py [hunk 3]
```diff
@@ -752,7 +752,7 @@ def numpy_to_vtk_matrix(array):
 
 
 def get_bounding_box_sizes(actor):
-    """Get the bounding box sizes of an actor."""
+    """Get the bounding box sizes of an actor"""
```
Comment (skoudoro) [inline fury/utils.py:755]: please, keep the dot

**Review by skoudoro — COMMENTED** [review]
ok, thanks for the info and update. 

I will try this PR a bit later and keep you updated if I have an issue. For now, See below my last comment. 

Thank you!

**Comment by skoudoro** [comment]
Also, you still have the same issue as we can see on the CI's. Let us know if you need help.

**Comment by antrikshmisri** [comment]
> Also, you still have the same issue as we can see on the CI's. Let us know if you need help.

Hey @skoudoro, I think the logic to distinguish between `UI` and `actor` needs some improvement. Instead of just checking for `add_to_scene`, we can check for multiple attributes like `add_to_scene` and `_setup`. In conjunction, we can confidently say if it is an actor or a UI. I tried to do this and all tests are passing. What do you think?

**Comment by antrikshmisri** [comment]
Hey @skoudoro, having `is_ui` method in `ui/helpers.py` causes circular imports 😕 . Logs can be seen [here](https://dev.azure.com/fury-gl/fury/_build/results?buildId=1573&view=logs&j=a846d25a-e32c-5640-1b53-e815fab94407&t=2f8d848e-c100-586c-cef4-f5504e3b245e&l=289).

File: fury/ui/tests/test_helpers.py [hunk 4]
```diff
@@ -1,9 +1,11 @@
 """Test helpers fonction ."""
 
+from fury.ui.containers import GridUI
+from fury.actor import Container
```
Comment (skoudoro) [inline fury/ui/tests/test_helpers.py:4]: Your circular is due to this which makes complete sense.

I recommend you create custom objects for your tests and not use `GridUI` or `Container`.

look [here](https://github.com/dipy/dipy/blob/master/dipy/workflows/tests/workflow_tests_utils.py) for example and their [dummy tests](https://github.com/dipy/dipy/blob/master/dipy/workflows/tests/test_iap.py#L79).

So, create custom class, inheriting from `ui.core.UI`

File: docs/tutorials/02_ui/viz_layout.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
+from fury.actor import square
+from fury.layout import GridLayout
+from fury import ui
+from fury import window
+import numpy as np
+from fury import actor
```
Comment (skoudoro) [inline docs/tutorials/02_ui/viz_layout.py:6]: please, update import order

**Review by skoudoro — COMMENTED** [review]
Hi @antrikshmisri,

See below for my comments

File: fury/ui/tests/test_helpers.py (hunk 4 above)
Comment (antrikshmisri) [reply fury/ui/tests/test_helpers.py:4]: Hey @skoudoro, besides this there is another circular import [here](https://github.com/fury-gl/fury/blob/8de602803381c33072775744db53c1329020d046/fury/layout.py#L5). Following the custom class track, I think a `DummyUI` class that inherits from `UI` should be created outside the `ui` sub-module, then the `is_ui` method will not be needed anymore and there will be no circular imports. Is this what you were implying?

Comment (skoudoro) [reply fury/ui/tests/test_helpers.py:4]: Please, see my PR and feel free to merge it.

File: docs/tutorials/02_ui/viz_layout.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
+from fury import ui
```
Comment (skoudoro) [inline docs/tutorials/02_ui/viz_layout.py:1]: You need to add a title for this example/tutorial and explain what is the goal of it

**Comment by antrikshmisri** [comment]
Hey @skoudoro, as said earlier the tests for `layout` are still failing because `is_ui` is being imported in it. Do I need to do something else here. Below is the Traceback:-
```python
fury\tests\test_layout.py:3: in <module>
    from fury import actor
fury\actor.py:12: in <module>
    from fury import layout
fury\layout.py:5: in <module>
    from fury.ui.helpers import is_ui
fury\ui\__init__.py:3: in <module>
    from fury.ui.containers import *
fury\ui\containers.py:12: in <module>
    from fury.actor import grid
E   ImportError: cannot import name 'grid'
```

File: docs/tutorials/02_ui/viz_layout.py (hunk 6 above)
Comment (antrikshmisri) [reply docs/tutorials/02_ui/viz_layout.py:12]: Done with latest commit.

File: fury/ui/helpers.py [hunk 7]
```diff
@@ -129,3 +123,15 @@ def check_overflow(textblock, width, overflow_postfix='',
             if side == 'left':
                 textblock.message = textblock.message[::-1]
             return mid_ptr
+
+
+def is_ui(actor):
```
Comment (skoudoro) [inline fury/ui/helpers.py:128]: you can put this in `fury.utils` for now.

**Review by skoudoro — COMMENTED** [review]
Hi @antrikshmisri,

Can you fix the conflict of this PR,  rebase it and move the function `is_ui`  to another module. See below. 

Thank you

File: fury/ui/helpers.py (hunk 7 above)
Comment (antrikshmisri) [reply fury/ui/helpers.py]: Done.

**Comment by antrikshmisri** [comment]
Hello @skoudoro, I have rebased the PR and also fixed the conflicts. Also, the `is_ui` function has been shifted to `fury.utils`

**Review by skoudoro — APPROVED** [review]
Thank you @antrikshmisri for the update. Looks good, merging

**Pull request merged**